Localization settings update

Skrypt opened the pull request:
Fixes an issue with default culture that could not be added to the supported cultures.
Also add the option to add InvariantCulture as a supported culture.

jtkech commented:
Based on our discussion i would have used.

            if (!String.IsNullOrEmpty(siteSettings.Culture))
            {
                options.SetDefaultCulture(siteSettings.Culture);
            }

            if (siteSettings.SupportedCultures.Length > 0)
            {
                var supportedCulture = new[] { options.DefaultRequestCulture.Culture.Name }
                    .Concat(siteSettings.SupportedCultures)
                    .Distinct()
                    .ToArray();

                options
                    .AddSupportedCultures(supportedCulture)
                    .AddSupportedUICultures(supportedCulture)
                    ;
            }

            app.UseRequestLocalization(options);


`options.DefaultRequestCulture` being the system one if our default culture has not been changed with one of our supported cultures. But i let you decide ;)

Skrypt commented:
I tested some things this afternoon and it's not right that we should only need to add the Default Culture when there's supported culture saved in the DB. I tried locally and my system culture is fr-CA and it doesn't get added to the supported cultures so I moved the code that adds the default culture to the supported culture out of that condition. Also `new[] { options.DefaultRequestCulture.Culture.Name }` fails when the default culture is InvariantCulture it won't create an array item with String.Empty.

Skrypt commented:
Alright I tried your code simplification and it seems to work. The code was different than what you did maybe the concat was not working properly.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Settings/Controllers/AdminController.cs:
```diff
@@ -124,18 +124,25 @@ public async Task<IActionResult> AddCulture(string systemCultureName, string cul
+
+                _notifier.Warning(H["The site needs to be restarted for the settings to take effect"]);
+            }
+            else {
```
formatting

Skrypt commented:
don't merge

jtkech commented:
> I tried locally and my system culture is fr-CA and it doesn't get added to the supported cultures so I moved the code that adds the default culture to the supported culture out of that condition.

Then you don't use the `supportedCulture` var if `siteSettings.SupportedCultures` is empty, so it's the same. 

But i don't trust you ;) here the default values of the `RequestLocalizationOptions`.

    private RequestCulture _defaultRequestCulture =
        new RequestCulture(CultureInfo.CurrentCulture, CultureInfo.CurrentUICulture);


    public IList<CultureInfo> SupportedCultures { get; set; } =
        new List<CultureInfo> { CultureInfo.CurrentCulture };

Skrypt commented:
The opposite scenario @jtkech.
I have 2 manually added cultures that are different than my system default culture.
The system culture doesn't get added if I put our new line in the if condition.

Skrypt commented:
Ok we fixed the remaining issues it is ready.

sebastienros commented:
Can you describe the scenarios it's handling?

sebastienros commented:
While you are on this topic, can you add a button from the "Add supported cultures" pages that would navigate to the previous settings page? I know you can click in the breadcrumb but it's not intuitive. Like a "Close" button

Skrypt commented:
The first issue we fixed is the fact that the default culture was not always added as a supported culture. So we tested the AcceptHeaderRequestCultureProvider and if the default culture is not a supported culture it will not work as intended.

Repro steps : You have en-US as a default culture with fr-FR as a supported culture. en-US will not be added as a supported culture. Add en-US in your browser settings and it won't translate the website in en-US. Though it will fallback to en-US if in your browser setting you have none of the supported cultures. Which in this case was only fr-FR.

Second issue was that InvariantCulture was never considered as a proper culture. So we could not add it as a supported culture and we could not set it manually as a default culture since the System Culture option had value="" which binds to null. So we now add value="culture.LCID" to make sure to save it as string.Empty only when the culture is invariant and null when we are using system culture.

Fixed the form that serves to add cultures to be able to also add InvariantCulture... and some others small fixes here and there to support it.

sebastienros commented:
Ok, waiting for @jtkech 's approval

Skrypt commented:
We just paired programmed for 1 hour

sebastienros commented:
In English ?

Skrypt commented:
We Skyped to fix this just half an hour ago.

Skrypt commented:
I'm adding the button

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Settings/Views/Admin/Culture.cshtml:
```diff
@@ -53,3 +63,4 @@
     </table>
 </fieldset>
 
+@Html.ActionLink(T["Close"].Value, "Index", "Admin", new { area = "OrchardCore.Settings", groupId = "general" }, new { @class = "btn btn-primary" })
```
`ActionLink(T["Close"].Value` no or there can be double encoding. Don't pass a translation to a helper that only accepts a string. Use the `<a>` tag directly in this case. and `ActionUrl`
